state_synchronizer/src/peer_manager.rs

```diff
@@ -4,24 +4,37 @@
 use crate::PeerId;
 use logger::prelude::*;
 use network::validator_network::StateSynchronizerSender;
-use rand::{thread_rng, Rng};
+use rand::{
+    distributions::{Distribution, WeightedIndex},
+    thread_rng,
+};
 use std::collections::HashMap;
 
+const MAX_SCORE: u64 = 100;
+
 #[derive(Default, Debug)]
 pub struct PeerInfo {
     is_alive: bool,
     is_upstream: bool,
+    score: u64,
 }
 
 impl PeerInfo {
-    pub fn new(is_alive: bool, is_upstream: bool) -> Self {
+    pub fn new(is_alive: bool, is_upstream: bool, score: u64) -> Self {
         Self {
             is_alive,
             is_upstream,
+            score,
         }
     }
 }
 
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum PeerScoreUpdateType {
+    Success,
+    InvalidChunk,
```

Comment:
nice. We should also handle timeout cases

Reply:
@phoenix-antigravity I'm having trouble determining where to dock off points on timeout. The only place I see is check_progress(), (https://github.com/libra/libra/blob/master/state_synchronizer/src/coordinator.rs#L352) but it seems to be a general check, not for a particular peer.

Reply:
So I added timeout check in request_next_chunk, based on the network error received from .send_to(). send_to() is called in process_chunk_response() as part of deliver_chunk() as well, should timeout also be checked there?